### PR TITLE
Refactor OIDC tests to better mimic an actual OIDC provider

### PR DESCRIPTION
**Pull request by sandhose**
This implements a fake OIDC server, which intercepts calls to the HTTP client.
It makes the tests more accurate, by covering more internal methods.

One particular example was the ID token validation, which was mocked before.
This helped me uncover a LIE we had in our dependencies: we actually need at least `authlib` 0.15.1, and not just >=0.14.0 (which looks fine for packagers, [according to repology](https://repology.org/project/python:authlib/versions))
I initially did this, because I needed to actually test the OIDC Backchannel Logout implementation in #11414

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -192,7 +192,7 @@ psycopg2 = { version = ">=2.8", markers = "platform_python_implementation != 'Py
 psycopg2cffi = { version = ">=2.8", markers = "platform_python_implementation == 'PyPy'", optional = true }
 psycopg2cffi-compat = { version = "==1.1", markers = "platform_python_implementation == 'PyPy'", optional = true }
 pysaml2 = { version = ">=4.5.0", optional = true }
-authlib = { version = ">=0.14.0", optional = true }
```
Comment (sandhose) [inline pyproject.toml:195]: So that was a LIE.

File: tests/federation/test_federation_client.py [hunk 2]
```diff
@@ -230,21 +227,3 @@ def _get_pdu_once(self) -> EventBase:
         self.assertEqual(remote_pdu.internal_metadata.outlier, False)
 
         return remote_pdu
-
-
-def _mock_response(resp: JsonDict):
```
Comment (sandhose) [inline tests/federation/test_federation_client.py:235]: Because we already had a `FakeResponse` class and I enhanced it, I took the opportunity to replace this here

File: tests/handlers/test_oidc.py [hunk 3]
```diff
@@ -175,18 +150,51 @@ def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
...
+        return patch.object(self.fake_provider, "get_metadata", return_value=metadata)
 
-        return patch.object(self.http_client, "get_json", patched_get_json)
+    def start_authorization(
+        self,
+        userinfo: dict,
+        client_redirect_url: str = "http://client/redirect",
+        scope: str = "openid",
+        with_sid: bool = False,
+    ) -> Tuple[SynapseRequest, FakeAuthorizationGrant]:
```
Comment (sandhose) [inline tests/handlers/test_oidc.py:183]: Even though we're not using the returned grant here, it is useful for the backchannel logout tests

File: tests/handlers/test_oidc.py [hunk 4]
```diff
@@ -424,48 +434,34 @@ def test_callback(self) -> None:
...
-        }
         username = "bar"
         userinfo = {
             "sub": "foo",
             "username": username,
         }
         expected_user_id = "@%s:%s" % (username, self.hs.hostname)
-        self.provider._exchange_code = simple_async_mock(return_value=token)  # type: ignore[assignment]
-        self.provider._parse_id_token = simple_async_mock(return_value=userinfo)  # type: ignore[assignment]
-        self.provider._fetch_userinfo = simple_async_mock(return_value=userinfo)  # type: ignore[assignment]
```
Comment (sandhose) [inline tests/handlers/test_oidc.py:440]: So all of those are not mocked anymore!

File: tests/test_utils/__init__.py [hunk 5]
```diff
@@ -97,27 +104,44 @@ async def cb(*args, **kwargs):
     return Mock(side_effect=cb)
 
 
-@attr.s
-class FakeResponse:
+# Type ignore: it does not fully implement IResponse, but is good enough for tests
+@zope.interface.implementer(IResponse)
```
Comment (sandhose) [inline tests/test_utils/__init__.py:108]: Not sure if that was the right way to do it, but `mypy` seems happy with that

File: tests/test_utils/__init__.py [hunk 6]
```diff
@@ -97,27 +104,44 @@ async def cb(*args, **kwargs):
...
 
     # body of the response
-    body = attr.ib(type=bytes)
+    body: bytes = b""
+
+    headers: Headers = Headers()
+
+    @property
+    def phrase(self):
+        return RESPONSES.get(self.code, b"Unknown Status")
```
Comment (sandhose) [inline tests/test_utils/__init__.py:129]: Instead of having to set an HTTP `phrase` explicitly, we just get it from the `RESPONSES` dict, which maps HTTP status codes to status phrases

File: tests/test_utils/oidc.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+            headers = Headers()
+
+        # Try to find the access token in the headers if any
+        auth_headers = headers.getRawHeaders(b"Authorization")
+        if auth_headers:
+            parts = auth_headers[0].split(b" ")
+            if parts[0] == b"Bearer" and len(parts) == 2:
+                access_token = parts[1].decode("ascii")
+
+        if method == "POST":
```
Comment (sandhose) [inline tests/test_utils/oidc.py:261]: One important thing to note here, is that this does not check for client auth, which I considered fine, since we have other tests covering this

File: synapse/handlers/oidc.py [hunk 8]
```diff
@@ -673,6 +674,12 @@ async def _parse_id_token(self, token: Token, nonce: str) -> CodeIDToken:
         Returns:
             The decoded claims in the ID token.
         """
+        id_token = token.get("id_token")
+
+        # That has been theoritically been checked by the caller, so even though
+        # assertion are not enabled in production, it is mainly here to appease mypy
+        assert id_token is not None
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/oidc.py:681]: We then fetch `id_token` again later on, which I don't think we need to do anymore.

See line 698 of the modified version.

File: synapse/handlers/oidc.py [hunk 9]
```diff
@@ -715,7 +722,9 @@ async def _parse_id_token(self, token: Token, nonce: str) -> CodeIDToken:
 
         logger.debug("Decoded id_token JWT %r; validating", claims)
 
-        claims.validate(leeway=120)  # allows 2 min of clock skew
+        claims.validate(
+            now=self._clock.time(), leeway=120
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/oidc.py:726]: Internally `validate` calls `now = time.time()` if `now` is not provided. Nice catch. 👍

File: tests/test_utils/oidc.py (hunk 7 above)
Comment (clokep) [reply tests/test_utils/oidc.py:261]: Maybe add a todo in the code here at least?

File: tests/test_utils/__init__.py [hunk 10]
```diff
@@ -97,27 +104,44 @@ async def cb(*args, **kwargs):
...
     there is a similar class at treq.test.test_response, but it lacks a `phrase`
     attribute, and didn't support deliverBody until recently.
     """
 
-    # HTTP response code
-    code = attr.ib(type=int)
+    verison: Tuple[bytes, int, int] = (b"HTTP", 1, 1)
 
-    # HTTP response phrase (eg b'OK' for a 200)
-    phrase = attr.ib(type=bytes)
```
Comment (clokep) [inline tests/test_utils/__init__.py:112]: Do we not use this? I'm quite confused by the changes in this file -- they seem unrelated to the rest of the PR.

Comment (clokep) [reply tests/test_utils/__init__.py:112]: Oh, I see this got moved below.

Still this refactoring seems like it could be split out for a separate review.

File: tests/test_utils/oidc.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+    userinfo: dict
+    client_id: str
+    redirect_uri: str
+    scope: str
+    nonce: Optional[str]
+    sid: Optional[str]
+
+
+class FakeOidcProvider:
+    """A fake OpenID Connect Provider."""
```
Comment (clokep) [inline tests/test_utils/oidc.py:44]: I'm not really sure what this is supposed to represent? Is this supposed to be a mocked version of `synapse.handlers.oidc.OidcProvider`? Can we say that explicitly? Or is it something more generic that is meant to be a "fake IdP" or something?

Comment (sandhose) [reply tests/test_utils/oidc.py]: I renamed this to `FakeOidcServer`, it should be clearer

File: tests/test_utils/oidc.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+    def reset_mocks(self):
+        self.request.reset_mock()
+        self.get_jwks_handler.reset_mock()
+        self.get_metadata_handler.reset_mock()
+        self.get_userinfo_handler.reset_mock()
+        self.post_token_handler.reset_mock()
+
+    def patch_homeserver(self, hs: HomeServer):
+        return patch.object(hs.get_proxied_http_client(), "request", self.request)
```
[the remark this answers is not in the document]
Comment (sandhose) [reply tests/test_utils/oidc.py]: I added a docstring. I'm not sure what makes more sense: even though we're only patching the HTTP client, it expects a HS as parameter, so that's why I named this `patch_homeserver` 🤷

File: tests/test_utils/oidc.py (hunk 7 above)
Comment (sandhose) [reply tests/test_utils/oidc.py:270]: Added a comment about that

**Comment by sandhose** [comment]
Looks like deb building in CI is broken because of an old version of authlib... even though the logs says it installs 1.1.0! And I can't reproduce the issue locally, `poetry run python ./scripts-dev/build_debian_packages.py` just works 🤔

File: synapse/handlers/oidc.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sandhose) [reply synapse/handlers/oidc.py:682]: Indeed. I removed that

File: tests/test_utils/__init__.py (hunk 5 above)
Comment (clokep) [reply tests/test_utils/__init__.py:108]: This is the correct way to do this! 👍

File: tests/test_utils/__init__.py [hunk 13]
```diff
@@ -97,27 +104,44 @@ async def cb(*args, **kwargs):
...
+    # HTTP response code
+    code: int = 200
 
     # body of the response
-    body = attr.ib(type=bytes)
+    body: bytes = b""
+
+    headers: Headers = Headers()
```
Comment (clokep) [inline tests/test_utils/__init__.py:125]: I think this needs to be a factory to avoid leaking between objects:

```suggestion
    headers: Headers = attr.Factory(Headers)
```

File: tests/test_utils/__init__.py [hunk 14]
```diff
@@ -97,27 +104,44 @@ async def cb(*args, **kwargs):
...
+    verison: Tuple[bytes, int, int] = (b"HTTP", 1, 1)
 
-    # HTTP response phrase (eg b'OK' for a 200)
-    phrase = attr.ib(type=bytes)
+    # HTTP response code
+    code: int = 200
 
     # body of the response
-    body = attr.ib(type=bytes)
+    body: bytes = b""
```
Comment (clokep) [inline tests/test_utils/__init__.py:123]: There previously was no default here, is this change needed?

File: tests/test_utils/oidc.py [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+        self.request = Mock(side_effect=self._request)
+        self.get_jwks_handler = Mock(side_effect=self._get_jwks_handler)
+        self.get_metadata_handler = Mock(side_effect=self._get_metadata_handler)
+        self.get_userinfo_handler = Mock(side_effect=self._get_userinfo_handler)
+        self.post_token_handler = Mock(side_effect=self._post_token_handler)
+
+        # A code -> grant mapping
+        self.authorization_grants: Dict[str, FakeAuthorizationGrant] = {}
+        # An access token -> grant mapping
+        self.sessions: Dict[str, FakeAuthorizationGrant] = {}
```
Comment (clokep) [inline tests/test_utils/oidc.py:69]: It looks like these two properties are private, can we prefix them with a `_`? (This applies to the other properties on this class too, but I think the others are public? Please double check.)

Additionally, I think `FakeAuthorizationGrant` is not meant to be seen outside of this module, I'd recommend renaming to `_FakeAuthorizationGrant`.

File: tests/handlers/test_oidc.py [hunk 16]
```diff
@@ -159,11 +134,11 @@ def default_config(self) -> Dict[str, Any]:
         return config
 
     def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
-        self.http_client = Mock(spec=["get_json"])
-        self.http_client.get_json.side_effect = get_json
-        self.http_client.user_agent = b"Synapse Test"
+        self.fake_provider = FakeOidcServer(clock=clock, issuer=ISSUER)
```
Comment (clokep) [inline tests/handlers/test_oidc.py:137]:
```suggestion
        self.fake_server = FakeOidcServer(clock=clock, issuer=ISSUER)
```

(and update the callers)

File: tests/handlers/test_oidc.py [hunk 17]
```diff
@@ -599,18 +577,22 @@ def test_callback_session(self) -> None:
     )
     def test_exchange_code(self) -> None:
         """Code exchange behaves correctly and handles various error scenarios."""
-        token = {"type": "bearer"}
-        token_json = json.dumps(token).encode("utf-8")
-        self.http_client.request = simple_async_mock(
-            return_value=FakeResponse(code=200, phrase=b"OK", body=token_json)
+        token = {
+            "type": "Bearer",
+            "access_token": "aabbcc",
```
Comment (clokep) [inline tests/handlers/test_oidc.py:582]: Is this access token just garbage or does it match with something?

File: tests/rest/client/test_auth.py [hunk 18]
```diff
@@ -465,9 +465,11 @@ def test_ui_auth_via_sso(self) -> None:
           * checking that the original operation succeeds
         """
 
+        fake_oidc_provider = self.helper.fake_oidc_server()
```
Comment (clokep) [inline tests/rest/client/test_auth.py:468]:
```suggestion
        fake_oidc_server = self.helper.fake_oidc_server()
```

(and all references -- there's also other instances in this file)

**Review by clokep — COMMENTED** [review]
I think this large looks good, most of my comments are fairly cursory.

File: tests/handlers/test_oidc.py (hunk 17 above)
Comment (sandhose) [reply tests/handlers/test_oidc.py:582]: It's garbage. We just check that `_exchange_code` returns the same dict later

File: tests/test_utils/__init__.py (hunk 14 above)
Comment (clokep) [reply tests/test_utils/__init__.py:123]: OK! Just noting it as a change in behavior. I think defaulting to no body isn't crazy.

File: tests/test_utils/oidc.py (hunk 15 above)
Comment (clokep) [reply tests/test_utils/oidc.py]: Arg, I meant to remove the second half of that when I realized it was used other places. Sorry about that. 👍

File: tests/rest/client/utils.py [hunk 19]
```diff
@@ -543,12 +553,28 @@ def upload_media(
...
+        """
+
+        return FakeOidcServer(
+            clock=self.hs.get_clock(),
+            issuer=issuer,
+        )
+
     def login_via_oidc(
         self,
+        provider: FakeOidcServer,
```
Comment (clokep) [inline tests/rest/client/utils.py:572]: Each of these `provider` parameters needs to be renamed to `server` or `fake_server`.

**Review by clokep — APPROVED** [review]
A couple minor renames, but looks good! 👍

File: tests/rest/client/utils.py (hunk 19 above)
Comment (sandhose) [reply tests/rest/client/utils.py]: Done!
